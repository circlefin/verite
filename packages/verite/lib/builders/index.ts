export * from "./common"
export * from "./input-descriptors"
export * from "./manifest-builder"
export * from "./output-descriptors"
export * from "./presentation-definition-builder"
