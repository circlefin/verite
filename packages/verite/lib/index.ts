export * from "./errors"
export * from "./issuer"
export * from "./submission-requests"
export * from "./utils"
export * from "./validators"
export * from "./verifier"
export * from "./builders"
export * from "./sample-data"
