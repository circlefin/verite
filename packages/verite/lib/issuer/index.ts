export * from "./credential-application"
export * from "./credential-offer"
export * from "./credential-fulfillment"
export * from "./revocation"
