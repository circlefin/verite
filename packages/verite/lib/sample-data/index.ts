export * from "./attestations"
export * from "./constants"
export * from "./manifests"
export * from "./presentation-definitions"
export * from "./verification-offer"
