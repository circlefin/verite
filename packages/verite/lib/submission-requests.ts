import fetch from "cross-fetch"
import { isString } from "lodash"
import { v4 as uuidv4 } from "uuid"

import type {
  ChallengeTokenUrlWrapper,
  VerificationOffer,
  SubmissionOffer,
  CredentialOffer
} from "../types"

const ONE_MONTH = 1000 * 60 * 60 * 24 * 30

/**
 * This file contains a set of functions to help interact with mobile wallets.
 * You can use any method of communication, but for the demo we decided to
 * have the mobile wallet scan a QR code.
 *
 * QR codes get more complex as you put more data in them. Consequently, the
 * QR code will encode a simple object with a `challengeTokenUrl` property that
 * references a URL. The client will scan the QR code and then follow the URL
 * provided.
 *
 * After following the URL, the client will receive a payload resembling the
 * schema as defined by the WACI protcol. https://identity.foundation/waci-presentation-exchange/#json-message
 * The actual data we wish to communicate to the client will be found in the
 * body property of the response. For example, it might contain the Credential
 * Manifest, a Presentation Request, a status URL for checking a verification,
 * or the challenge string necessary for signing presentations.
 *
 * Most importantly, we have made some modifications, such as adding a
 * `replyUrl` to instruct the client where to submit their credential
 * application.
 */

/**
 * Build a wrapper containing a challenge token URL for use in a QR code.
 */
export function challengeTokenUrlWrapper(
  challengeTokenUrl: string
): ChallengeTokenUrlWrapper {
  return {
    challengeTokenUrl
  }
}

export function buildRequestCommon(
  id: string,
  type: string,
  from: string,
  replyUrl: string,
  statusUrl?: string
): SubmissionOffer {
  const now = Date.now()
  const expires = now + ONE_MONTH

  const result = {
    id,
    type,
    from,
    created_time: new Date(now).toISOString(),
    expires_time: new Date(expires).toISOString(),
    reply_url: replyUrl,
    body: {
      status_url: statusUrl,
      challenge: uuidv4()
    }
  }
  return result
}

/**
 * Handle QR codes that initiate an issuance or verification workflow.
 *
 * When scanning a QR code, it will encode a JSON object with a challengeTokenUrl
 * property. We will subsequently fetch that value to retrieve the full
 * manifest or verification request document.
 */
export async function handleScan(
  scanData: string
): Promise<CredentialOffer | VerificationOffer | undefined> {
  const payload = json(scanData) as ChallengeTokenUrlWrapper

  if (!payload.challengeTokenUrl) {
    return
  }

  const response = await fetch(payload.challengeTokenUrl)
  return response.json()
}

function json(body: string | Record<string, unknown>): Record<string, unknown> {
  if (isString(body)) {
    try {
      return JSON.parse(body)
    } catch (e) {
      return {}
    }
  }

  return body
}
