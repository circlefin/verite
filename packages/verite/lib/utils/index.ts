export * from "./async-fns"
export * from "./attestation-registry"
export * from "./bitstring"
export * from "./claim-format"
export * from "./constants"
export * from "./credentials"
export * from "./did-fns"
export * from "./has-paths"
