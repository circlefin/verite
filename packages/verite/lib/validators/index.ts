export * from "./validate-presentation-submission"
export * from "./validate-credential-application"
