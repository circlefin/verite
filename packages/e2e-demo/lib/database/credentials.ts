import { BitBuffer } from "bit-buffers"
import { compact, random, sample } from "lodash"
import { v4 as uuidv4 } from "uuid"
import {
  isRevocable,
  MaybeRevocableCredential,
  RevocableCredential,
  StatusList2021Entry,
  StatusList2021Credential,
  MINIMUM_BITSTREAM_LENGTH,
  asyncMap,
  decodeVerifiableCredential,
  generateRevocationList,
  buildIssuer,
  EncodedStatusListCredential,
  generateEncodedRevocationList
} from "verite"

import { fullURL } from "../utils"
import { prisma, User, Credential } from "./prisma"
import { findUser } from "./users"

export type DecodedDatabaseCredential<T = MaybeRevocableCredential> =
  Credential & {
    credential: T
  }

/**
 *
 */
export const storeCredentials = async (
  credentials: Array<MaybeRevocableCredential>,
  userId: string
): Promise<void> => {
  await asyncMap(credentials, async (credential) => {
    await prisma.credential.create({
      data: {
        userId,
        jwt: credential.proof.jwt
      }
    })
  })
}

/**
 * Returns all revocation lists, optionally filtered by ones for a specific
 * host.
 *
 * @param host to filter by
 * @returns revocation lists filtered by host
 */
export const findAllOrCreateRevocationLists = async (): Promise<
  StatusList2021Credential[]
> => {
  const lists = await prisma.revocationList.findMany()

  // Create one if it doesn't exist.
  if (lists.length === 0) {
    const list = await generateRevocationList({
      statusList: [],
      url: fullURL(`/api/demos/revocation/${uuidv4()}`),
      issuer: process.env.ISSUER_DID,
      signer: buildIssuer(process.env.ISSUER_DID, process.env.ISSUER_SECRET)
    })
    await saveRevocationList(list)
    return [list]
  }

  return await asyncMap(lists, async (list) => {
    const bits = BitBuffer.fromBitstring(list.encodedList)

    return generateRevocationList({
      statusList: bits.toIndexArray(),
      url: fullURL(`/api/demos/revocation/${list.id}`),
      issuer: process.env.ISSUER_DID,
      signer: buildIssuer(process.env.ISSUER_DID, process.env.ISSUER_SECRET)
    })
  })
}

/**
 *
 */
export const findRevocationListForCredential = async (
  credential: RevocableCredential
): Promise<StatusList2021Credential | undefined> => {
  if (!credential.credentialStatus) {
    return
  }

  const url = credential.credentialStatus.statusListCredential
  const revocationLists = await findAllOrCreateRevocationLists()
  return revocationLists.find((list) => list.id === url)
}

/**
 *
 */
export const findUserByCredential = async (
  jwt: string
): Promise<User | undefined> => {
  const data = await prisma.credential.findFirst({
    where: { jwt }
  })
  if (data) {
    const userId = data.userId
    const user = await findUser(userId)
    return user
  }
}

/**
 *
 */
export const findCredentialsByUserId = async (
  userId: string
): Promise<DecodedDatabaseCredential[]> => {
  const records = await prisma.credential.findMany({
    where: {
      userId
    }
  })

  return decodeDatabaseCredentials(records)
}

/**
 *
 */
export const findCredentialsByUserIdAndType = async (
  userId: string,
  type: string
): Promise<MaybeRevocableCredential[]> => {
  const credentials = (await findCredentialsByUserId(userId))
    .filter((credential) => credential.credential.type[1] === type)
    .map((c) => c.credential)

  return credentials
}

/**
 * Finds the newest credential since a given date. Useful for demo purposes.
 *
 * @param createdAt will restrict search to only credentials older than this date, defaults to epoch
 * @returns a decoded credential
 */
export const findNewestCredentialSinceDate = async (
  createdAt: Date,
  userId?: string
): Promise<MaybeRevocableCredential | undefined> => {
  const result = await prisma.credential.findFirst({
    where: {
      createdAt: {
        gt: createdAt
      },
      userId: userId
    },
    orderBy: {
      createdAt: "desc"
    }
  })

  if (result) {
    try {
      const decoded = await decodeVerifiableCredential(result.jwt)
      return decoded
    } catch (e) {
      return
    }
  }
}

/**
 *
 */
const findCredentialsByRevocationlist = async (
  revocationList: StatusList2021Credential
): Promise<DecodedDatabaseCredential<RevocableCredential>[]> => {
  // NOTE: This is not efficient. In production you would normally have an index
  // to allow for these credentials to be mapped to a revocation list record.
  const records = await prisma.credential.findMany()
  const decoded = await decodeDatabaseCredentials(records)

  // Remove non-revocable credentials
  const revocable = decoded.filter(({ credential }) =>
    isRevocable(credential)
  ) as DecodedDatabaseCredential<RevocableCredential>[]

  return revocable.filter(
    ({ credential }) =>
      credential.credentialStatus.statusListCredential.split("/").pop() ===
      revocationList.id
  )
}

/**
 * Find a revocation list by its ID
 */
export const getRevocationListById = async (
  id: string
): Promise<EncodedStatusListCredential | undefined> => {
  const revocationList = await prisma.revocationList.findFirst({
    where: {
      id
    }
  })

  if (!revocationList) {
    return
  }

  const bits = BitBuffer.fromBitstring(revocationList.encodedList)

  return generateEncodedRevocationList({
    statusList: bits.toIndexArray(),
    url: fullURL(`/api/demos/revocation/${revocationList.id}`),
    issuer: process.env.ISSUER_DID,
    signer: buildIssuer(process.env.ISSUER_DID, process.env.ISSUER_SECRET)
  })
}

/**
 * Upsert a Revocation List record to the database
 */
export const saveRevocationList = async (
  revocationList: StatusList2021Credential
): Promise<void> => {
  // Use the uuid from the URL as the ID
  const id = revocationList.id.split("/").pop()

  await prisma.revocationList.upsert({
    where: {
      id
    },
    create: {
      id,
      encodedList: revocationList.credentialSubject.encodedList
    },
    update: {
      id,
      encodedList: revocationList.credentialSubject.encodedList
    }
  })
}

/**
 * Select a random revocation list and a random index from that list
 * to store the revocation status.
 *
 * Each revocable credential requires that we provide it a unique index in a list.
 *
 * @returns a revocation list status containing a list and index
 */
export const generateRevocationListStatus =
  async (): Promise<StatusList2021Entry> => {
    // Pick a random revocation list
    const lists = await findAllOrCreateRevocationLists()
    const revocationList = sample(lists)

    // Find all credentials in the revocation list and map the index
    const consumedIndexes = (
      await findCredentialsByRevocationlist(revocationList)
    ).map(({ credential }) =>
      parseInt(credential.credentialStatus.statusListIndex, 10)
    )

    // Try up to 10 times for now
    for (let i = 0; i < MINIMUM_BITSTREAM_LENGTH; i++) {
      const randomIndex = random(0, MINIMUM_BITSTREAM_LENGTH - 1)
      const index = consumedIndexes.indexOf(randomIndex)
      if (index === -1) {
        return {
          id: `${revocationList.id}#${randomIndex}`,
          type: "StatusList2021Entry",
          statusPurpose: "revocation",
          statusListIndex: randomIndex.toString(),
          statusListCredential: revocationList.id
        }
      }
    }
  }

async function decodeDatabaseCredentials(
  records: Credential[]
): Promise<DecodedDatabaseCredential[]> {
  const credentials = await asyncMap(records, async (record) => {
    try {
      const decoded = await decodeVerifiableCredential(record.jwt)

      return {
        ...record,
        credential: decoded
      }
    } catch (e) {}
  })

  return compact(credentials)
}
