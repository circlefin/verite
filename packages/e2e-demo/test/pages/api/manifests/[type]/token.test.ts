import { temporaryAuthToken } from "../../../../../lib/database"
import handler from "../../../../../pages/api/demos/issuer/manifests/[type]/[token]"
import { userFactory } from "../../../../factories"
import { createMocks } from "../../../../support/mocks"

describe("GET /api/demos/issuer/manifests/[type]/[token]", () => {
  it("returns the KYC/AML manifest", async () => {
    expect.assertions(4)

    const user = await userFactory()
    const token = await temporaryAuthToken(user)

    const { req, res } = createMocks({
      method: "GET",
      query: { type: "kyc", token }
    })

    await handler(req, res)

    const json = res._getJSONData()
    const manifest = json.body.manifest
    expect(res.statusCode).toBe(200)
    expect(json.reply_url).toBeDefined()
    expect(manifest.id).toEqual("KYCAMLManifest")
    expect(manifest).toMatchSnapshot()
  })

  it("returns the CreditScore manifest", async () => {
    expect.assertions(4)

    const user = await userFactory()
    const token = await temporaryAuthToken(user)
    const { req, res } = createMocks({
      method: "GET",
      query: { type: "credit-score", token }
    })

    await handler(req, res)

    const json = res._getJSONData()
    const manifest = json.body.manifest
    expect(res.statusCode).toBe(200)
    expect(json.reply_url).toBeDefined()
    expect(manifest.id).toEqual("CreditScoreManifest")
    expect(manifest).toMatchSnapshot()
  })

  it("returns 404 if no manifest is found", async () => {
    const { req, res } = createMocks({
      method: "GET",
      query: { type: "invalid", token: "something-invalid" }
    })

    await handler(req, res)

    expect(res.statusCode).toBe(404)
  })
})
