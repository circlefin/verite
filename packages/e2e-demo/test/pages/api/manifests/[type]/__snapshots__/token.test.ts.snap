// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /api/demos/issuer/manifests/[type]/[token] returns the CreditScore manifest 1`] = `
Object {
  "format": Object {
    "jwt_vc": Object {
      "alg": Array [
        "EdDSA",
      ],
    },
    "jwt_vp": Object {
      "alg": Array [
        "EdDSA",
      ],
    },
  },
  "id": "CreditScoreManifest",
  "issuer": Object {
    "id": "did:web:demos.verite.id",
    "name": "Verite",
    "styles": Object {},
  },
  "output_descriptors": Array [
    Object {
      "description": "Attestation that Verite has performed a Credit Score check for this subject",
      "display": Object {
        "description": Object {
          "text": "The Credit Score authority processes credit worthiness analysis, potentially employing a number of internal and external vendor providers.",
        },
        "properties": Array [
          Object {
            "label": "Score",
            "path": Array [
              "$.CreditScoreAttestation.score",
            ],
            "schema": Object {
              "type": "number",
            },
          },
          Object {
            "label": "Score Type",
            "path": Array [
              "$.CreditScoreAttestation.scoreType",
            ],
            "schema": Object {
              "type": "string",
            },
          },
          Object {
            "label": "Provider",
            "path": Array [
              "$.CreditScoreAttestation.provider",
            ],
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "subtitle": Object {
          "fallback": "Includes credit score",
          "path": Array [
            "$.CreditScoreAttestation.scoreType",
          ],
        },
        "title": Object {
          "text": "Verite Risk Score",
        },
      },
      "id": "CreditScoreCredential",
      "name": "Proof of Credit Score from Verite",
      "schema": "https://verite.id/definitions/schemas/0.0.1/CreditScoreAttestation",
      "styles": Object {
        "background": Object {
          "color": "#8B5CF6",
        },
        "hero": Object {
          "alt": "Credit Score Visual",
          "uri": "http://test.host/img/credit-score-hero.png",
        },
        "text": Object {
          "color": "#FFFFFF",
        },
        "thumbnail": Object {
          "alt": "Verite Logo",
          "uri": "http://test.host/img/credit-score-thumbnail.png",
        },
      },
    },
  ],
  "presentation_definition": Object {
    "format": Object {
      "jwt_vp": Object {
        "alg": Array [
          "EdDSA",
        ],
      },
    },
    "id": "ProofOfControlPresentationDefinition",
    "input_descriptors": Array [
      Object {
        "constraints": Object {
          "fields": Array [
            Object {
              "id": "holder",
              "path": Array [
                "$.holder",
              ],
              "purpose": "The VP should contain a DID in the holder, which is the same DID that signs the VP. This DID will be used as the subject of the issued VC",
            },
          ],
        },
        "format": Object {
          "jwt_vp": Object {
            "alg": Array [
              "EdDSA",
            ],
          },
        },
        "id": "proofOfIdentifierControlVP",
        "name": "Proof of Control Verifiable Presentation",
        "purpose": "A VP establishing proof of identifier control over the DID.",
      },
    ],
  },
  "spec_version": "https://identity.foundation/credential-manifest/spec/v1.0.0/",
}
`;

exports[`GET /api/demos/issuer/manifests/[type]/[token] returns the KYC/AML manifest 1`] = `
Object {
  "format": Object {
    "jwt_vc": Object {
      "alg": Array [
        "EdDSA",
      ],
    },
    "jwt_vp": Object {
      "alg": Array [
        "EdDSA",
      ],
    },
  },
  "id": "KYCAMLManifest",
  "issuer": Object {
    "id": "did:web:demos.verite.id",
    "name": "Verite",
    "styles": Object {},
  },
  "output_descriptors": Array [
    Object {
      "description": "Attestation that Verite has completed KYCAMLAttestation verification for this subject",
      "display": Object {
        "description": Object {
          "text": "The issuing authority processes KYCAMLAttestation analysis, potentially employing a number of internal and external vendor providers.",
        },
        "properties": Array [
          Object {
            "label": "Process",
            "path": Array [
              "$.KYCAMLAttestation.process",
            ],
            "schema": Object {
              "type": "string",
            },
          },
          Object {
            "label": "Approved At",
            "path": Array [
              "$.KYCAMLAttestation.approvalDate",
            ],
            "schema": Object {
              "format": "date-time",
              "type": "string",
            },
          },
        ],
        "subtitle": Object {
          "fallback": "Includes date of approval",
          "path": Array [
            "$.approvalDate",
            "$.vc.approvalDate",
          ],
        },
        "title": Object {
          "text": "Verite KYCAMLAttestation",
        },
      },
      "id": "KYCAMLCredential",
      "name": "KYCAMLCredential from Verite",
      "schema": "https://verite.id/definitions/schemas/0.0.1/KYCAMLAttestation",
      "styles": Object {
        "background": Object {
          "color": "#EC4899",
        },
        "hero": Object {
          "alt": "KYC+AML Visual",
          "uri": "http://test.host/img/kyc-aml-hero.png",
        },
        "text": Object {
          "color": "#FFFFFF",
        },
        "thumbnail": Object {
          "alt": "Verite Logo",
          "uri": "http://test.host/img/kyc-aml-thumbnail.png",
        },
      },
    },
  ],
  "presentation_definition": Object {
    "format": Object {
      "jwt_vp": Object {
        "alg": Array [
          "EdDSA",
        ],
      },
    },
    "id": "ProofOfControlPresentationDefinition",
    "input_descriptors": Array [
      Object {
        "constraints": Object {
          "fields": Array [
            Object {
              "id": "holder",
              "path": Array [
                "$.holder",
              ],
              "purpose": "The VP should contain a DID in the holder, which is the same DID that signs the VP. This DID will be used as the subject of the issued VC",
            },
          ],
        },
        "format": Object {
          "jwt_vp": Object {
            "alg": Array [
              "EdDSA",
            ],
          },
        },
        "id": "proofOfIdentifierControlVP",
        "name": "Proof of Control Verifiable Presentation",
        "purpose": "A VP establishing proof of identifier control over the DID.",
      },
    ],
  },
  "spec_version": "https://identity.foundation/credential-manifest/spec/v1.0.0/",
}
`;
